app/src/org/commcare/android/framework/StateFragment.java

```diff
@@ -10,95 +10,94 @@
+    }
+
+    @Override
+    public void onAttach(Activity activity) {
         super.onAttach(activity);
-        if(activity instanceof CommCareActivity) {
+
+        if (activity instanceof CommCareActivity) {
             this.boundActivity = (CommCareActivity)activity;
             this.boundActivity.stateHolder = this;
-            if(this.currentTask != null && this.currentTask.getStatus() == AsyncTask.Status.RUNNING) {
+
+            if (isCurrentTaskRunning()) {
                 this.currentTask.connect(boundActivity);
             }
         }
-      }
-      
-      public CommCareActivity getPreviousState() { 
-          return lastActivity;
-      }
-      
-      public void connectTask(CommCareTask task) {
-          this.currentTask = task;
-      }
-     
-      /**
-       * This method will only be called once when the retained
-       * Fragment is first created.
-       */
-      @Override
-      public void onCreate(Bundle savedInstanceState) {
-        super.onCreate(savedInstanceState);
-     
-        // Retain this fragment across configuration changes.
-        setRetainInstance(true);
-      }
-     
-      /**
-       * Set the callback to null so we don't accidentally leak the 
-       * Activity instance.
-       */
-      @Override
-      public void onDetach() {
+    }
+
+    private boolean isCurrentTaskRunning() {
+        return this.currentTask != null &&
+                this.currentTask.getStatus() == AsyncTask.Status.RUNNING;
+    }
+
+    @Override
+    public void onDetach() {
         super.onDetach();
-        if(this.boundActivity != null) {
+
+        if (this.boundActivity != null) {
             lastActivity = boundActivity;
         }
-        if(currentTask != null) {
+
+        if (currentTask != null) {
             Log.i("CommCareUI", "Detaching activity from current task: " + this.currentTask);
             currentTask.disconnect();
-            unlock();
+            releaseWakeLock();
         }
-      }
+    }
 
     public void cancelTask() {
-        if(currentTask != null) {
+        if (currentTask != null) {
             currentTask.cancel(false);
         }
     }
 
-    private WakeLock wakelock;
-
-    public synchronized void wakelock(int lockLevel) {
-        if(wakelock != null) {
-            if(wakelock.isHeld()) {
+    public synchronized void acquireWakeLock(int lockLevel) {
+        if (wakelock != null) {
+            if (wakelock.isHeld()) {
                 wakelock.release();
             }
         }
-        PowerManager pm = (PowerManager) boundActivity.getSystemService(Context.POWER_SERVICE);
+
+        PowerManager pm = (PowerManager)boundActivity.getSystemService(Context.POWER_SERVICE);
         wakelock = pm.newWakeLock(lockLevel, "CommCareLock");
         wakelock.acquire();
     }
 
-    public synchronized void unlock() {
-        if(wakelock != null) {
+    public synchronized void releaseWakeLock() {
+        if (wakelock != null) {

```

Comment:
Should this also check for wakelock.isHeld(), since you are doing that before you call release() in acquireWakeLock?